### PR TITLE
add tests for pop and protected path

### PR DESCRIPTION
**Pull request by andriidovhan**

### CODE REVIEW

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy [hunk 1]
```diff
@@ -1183,6 +1184,68 @@ class ProtectedPathV2Spec extends HealthCheckSpecification {
...
+        def flow = flowHelperV2.randomFlow(swPair)
+        flow.allocateProtectedPath = true
+        flowHelperV2.addFlow(flow)
+
+        then: "Main path is built through the preferable path(tr1 or tr2)"
+        def flowPaths = northbound.getFlowPath(flow.flowId)
+        def realFlowPathInvolvedSwitches = pathHelper.getInvolvedSwitches(pathHelper.convert(flowPaths))*.dpId
+        realFlowPathInvolvedSwitches == involvedSwP1 || realFlowPathInvolvedSwitches == involvedSwP2
+
+        and: "Protected path is built through the non preferable path(tr3)"
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy]: can you elaborate here please? I mean the reason why it picked 'tr3' path. It is really hard to follow all this pop logic without any pictures before eyes and with very brief comments. I mean I get it right now, but it will take me too much time to refresh the memory in a month or two. And the test must be a good helper here

Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy]: left the comment

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy [hunk 2]
```diff
@@ -1183,6 +1184,68 @@ class ProtectedPathV2Spec extends HealthCheckSpecification {
...
+        assumeTrue("Not enough amount of diverse paths", allPathCandidates.size() >= 3)
+        //select paths for further manipulations
+        def mainPath1 = allPathCandidates.first()
+        def mainPath2 = allPathCandidates.find { it != mainPath1 }
+        def protectedPath = allPathCandidates.find { it != mainPath1 && it != mainPath2}
+        def involvedSwP1 = pathHelper.getInvolvedSwitches(mainPath1)*.dpId
+        def involvedSwP2 = pathHelper.getInvolvedSwitches(mainPath2)*.dpId
+        def involvedSwProtected = pathHelper.getInvolvedSwitches(protectedPath)*.dpId
+
+        and: "Src, dst and transit switches belongs to different POPs(src:1, dst:4, tr1/tr2:2, tr3:3)"
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy]: you are right

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy [hunk 3]
```diff
@@ -1183,6 +1184,68 @@ class ProtectedPathV2Spec extends HealthCheckSpecification {
...
+                p1.intersect(p2) == [] ? 1 : 0
+            }.size() >= 3
+        } ?: assumeTrue("No suiting switches found", false)
+
+        //find out all paths with 3 switches
+        def allPaths = swPair.paths.findAll { pathHelper.getInvolvedSwitches(it).size() == 3 }
+        def allPathCandidates = allPaths.unique(false) { a, b ->
+            a.intersect(b) == [] ? 1 : 0
+        }
+        assumeTrue("Not enough amount of diverse paths", allPathCandidates.size() >= 3)
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy]: it was refactored

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy [hunk 4]
```diff
@@ -1183,6 +1184,68 @@ class ProtectedPathV2Spec extends HealthCheckSpecification {
...
+        database.setSwitchPop(swPair.dst.dpId, "3")
+        database.setSwitchPop(involvedSwProtected[1], "4")
+
+        and: "Path which contains tr3 is non preferable"
+        allPaths.findAll { involvedSwProtected[1] in pathHelper.getInvolvedSwitches(it)*.dpId }.each {
+            pathHelper.getInvolvedIsls(it).each {
+                northbound.updateLinkProps([islUtils.toLinkProps(it,
+                        ["cost": (database.getIslCost(it) + 10).toString()])])
+            }
+        }
```
[the remark this answers is not in the document]
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy]: why? I mean how do you understand this is not a defect?

[the remark this answers is not in the document]
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy]: fixed, it should be clear now

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy [hunk 5]
```diff
@@ -1303,6 +1304,106 @@ class ProtectedPathV2Spec extends HealthCheckSpecification {
...
+        List<Isl> islsToUpdate = []
+        allPaths.findAll { involvedSwProtected[1] in pathHelper.getInvolvedSwitches(it)*.dpId }.each {
+            pathHelper.getInvolvedIsls(it).each {
+                islsToUpdate << it
+            }
+        }
+        def involvedIslsOfMainPath = pathHelper.getInvolvedIsls(mainPath1)
+        def totalCostOfMainPath = involvedIslsOfMainPath.sum { northbound.getLink(it).cost ?: 700 }
+        def amountOfIlslsOnMainPath = involvedIslsOfMainPath.size()
+        Integer newIslCost = ((totalCostOfMainPath + (amountOfIlslsOnMainPath * 2000) - 1) /
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy]: the comment explains 2000, but I think it still makes sense to extract it into variables and make a clean self-explaining code `totalCostOfMainPath + (amountOfInvolvedIsls * diversity.pop.isl.cost * diversityGroupPerPopUseCounter) - 1`

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy [hunk 6]
```diff
@@ -1303,6 +1304,106 @@ class ProtectedPathV2Spec extends HealthCheckSpecification {
...
+         *
+         *             srcSwitch_POP_1
+         *             /       |       \
+         *            / 700    | 700    \ newIslCost
+         *           /         |         \
+         *   trSw1_POP_2   trSw2_POP_2   trSw3_POP_3
+         *          \          |         /
+         *           \ 700     | 700    / newIslCost
+         *            \        |       /
+         *             dstSwitch_POP_4
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy]: :+1:

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy [hunk 7]
```diff
@@ -1303,6 +1304,106 @@ class ProtectedPathV2Spec extends HealthCheckSpecification {
...
+         *  diversityGroupPerPopUseCounter = amount of switches in the same POP
+         *  (in this test there are two switches in the same POP) */
+        List<Isl> islsToUpdate = []
+        allPaths.findAll { involvedSwProtected[1] in pathHelper.getInvolvedSwitches(it)*.dpId }.each {
+            pathHelper.getInvolvedIsls(it).each {
+                islsToUpdate << it
+            }
+        }
+        def involvedIslsOfMainPath = pathHelper.getInvolvedIsls(mainPath1)
+        def totalCostOfMainPath = involvedIslsOfMainPath.sum { northbound.getLink(it).cost ?: 700 }
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy]: 1. done
2. I would like to avoid `assume` in the test. Actually there are only two requests(`northbound.getLink`), so I guess it is not critical to leave it as is.

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy (hunk 5 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy]: it makes sense to me.
done.
